File: java/src/main/java/com/squirrelparser/MetaGrammar.java
package com.squirrelparser;

import static com.squirrelparser.Utils.unescapeChar;
import static com.squirrelparser.Utils.unescapeString;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * MetaGrammar: A grammar for defining PEG grammars.
 */
public final class MetaGrammar {
    private static final String TERMINAL_LABEL = "<Terminal>";

    /** The meta-grammar rules for parsing PEG grammars. */
    public static final Map<String, Clause> RULES;

    static {
        RULES = new HashMap<>();
        RULES.put("Grammar", new Seq(List.of(
            new Ref("WS"),
            new OneOrMore(new Seq(List.of(
                new Ref("Rule"),
                new Ref("WS")
            )))
        )));
        RULES.put("Rule", new Seq(List.of(
            new Optional(new Str("~")),
            new Ref("Identifier"),
            new Ref("WS"),
            new Str("<-"),
            new Ref("WS"),
            new Ref("Expression"),
            new Ref("WS"),
            new Str(";"),
            new Ref("WS")
        )));
        RULES.put("Expression", new Ref("Choice"));
        RULES.put("Choice", new Seq(List.of(
            new Ref("Sequence"),
            new ZeroOrMore(new Seq(List.of(
                new Ref("WS"),
                new Str("/"),
                new Ref("WS"),
                new Ref("Sequence")
            )))
        )));
        RULES.put("Sequence", new Seq(List.of(
            new Ref("Prefix"),
            new ZeroOrMore(new Seq(List.of(
                new Ref("WS"),
                new Ref("Prefix")
            )))
        )));
        RULES.put("Prefix", new First(List.of(
            new Seq(List.of(new Str("&"), new Ref("WS"), new Ref("Prefix"))),
            new Seq(List.of(new Str("!"), new Ref("WS"), new Ref("Prefix"))),
            new Seq(List.of(new Str("~"), new Ref("WS"), new Ref("Prefix"))),
            new Ref("Suffix")
        )));
        RULES.put("Suffix", new First(List.of(
            new Seq(List.of(new Ref("Suffix"), new Ref("WS"), new Str("*"))),
            new Seq(List.of(new Ref("Suffix"), new Ref("WS"), new Str("+"))),
            new Seq(List.of(new Ref("Suffix"), new Ref("WS"), new Str("?"))),
            new Ref("Primary")
        )));
        RULES.put("Primary", new First(List.of(
            new Ref("Identifier"),
            new Ref("StringLiteral"),
            new Ref("CharLiteral"),
            new Ref("CharClass"),
            new Ref("AnyChar"),
            new Ref("Parens")
        )));
        RULES.put("Parens", new Seq(List.of(
            new Str("("),
            new Ref("WS"),
            new Optional(new Ref("Expression")),
            new Ref("WS"),
            new Str(")")
        )));
        RULES.put("Identifier", new Seq(List.of(
            new First(List.of(
                CharSet.range("a", "z"),
                CharSet.range("A", "Z"),
                new Char("_")
            )),
            new ZeroOrMore(new First(List.of(
                CharSet.range("a", "z"),
                CharSet.range("A", "Z"),
                CharSet.range("0", "9"),
                new Char("_")
            )))
        )));
        RULES.put("StringLiteral", new Seq(List.of(
            new Str("\""),
            new ZeroOrMore(new First(List.of(
                new Ref("EscapeSequence"),
                new Seq(List.of(
                    new NotFollowedBy(new First(List.of(new Str("\""), new Str("\\")))),
                    new AnyChar()
                ))
            ))),
            new Str("\"")
        )));
        RULES.put("CharLiteral", new Seq(List.of(
            new Str("'"),
            new First(List.of(
                new Ref("EscapeSequence"),
                new Seq(List.of(
                    new NotFollowedBy(new First(List.of(new Str("'"), new Str("\\")))),
                    new AnyChar()
                ))
            )),
            new Str("'")
        )));
        RULES.put("EscapeSequence", new Seq(List.of(
            new Str("\\"),
            new First(List.of(
                new Char("n"),
                new Char("r"),
                new Char("t"),
                new Char("\\"),
                new Char("\""),
                new Char("'"),
                new Char("["),
                new Char("]"),
                new Char("-")
            ))
        )));
        RULES.put("CharClass", new Seq(List.of(
            new Str("["),
            new Optional(new Str("^")),
            new OneOrMore(new First(List.of(
                new Ref("CharRange"),
                new Ref("CharClassChar")
            ))),
            new Str("]")
        )));
        RULES.put("CharRange", new Seq(List.of(
            new Ref("CharClassChar"),
            new Str("-"),
            new Ref("CharClassChar")
        )));
        RULES.put("CharClassChar", new First(List.of(
            new Ref("EscapeSequence"),
            new Seq(List.of(
                new NotFollowedBy(new First(List.of(new Str("]"), new Str("\\"), new Str("-")))),
                new AnyChar()
            ))
        )));
        RULES.put("AnyChar", new Str("."));
        RULES.put("~WS", new ZeroOrMore(new First(List.of(
            new Char(" "),
            new Char("\t"),
            new Char("\n"),
            new Char("\r"),
            new Ref("Comment")
        ))));
        RULES.put("Comment", new Seq(List.of(
            new Str("#"),
            new ZeroOrMore(new Seq(List.of(
                new NotFollowedBy(new Char("\n")),
                new AnyChar()
            ))),
            new Optional(new Char("\n"))
        )));
    }

    /**
     * Parse a grammar specification and return the rules.
     */
    public static Map<String, Clause> parseGrammar(String grammarSpec) {
        var parser = new Parser(RULES, "Grammar", grammarSpec);
        var parseResult = parser.parse();
        if (parseResult.hasSyntaxErrors()) {
            var errors = parseResult.getSyntaxErrors().stream()
                .map(Object::toString)
                .toList();
            throw new IllegalArgumentException("Failed to parse grammar. Syntax errors:\n" +
                String.join("\n", errors));
        }

        var ast = ASTBuilder.buildAST(parseResult);
        var grammarMap = new HashMap<String, Clause>();

        for (var ruleNode : ast.children()) {
            if (!ruleNode.label().equals("Rule")) {
                continue;
            }

            ASTNode identifierNode = null;
            for (var child : ruleNode.children()) {
                if (child.label().equals("Identifier")) {
                    identifierNode = child;
                    break;
                }
            }
            if (identifierNode == null) {
                throw new IllegalArgumentException("Rule has no Identifier child");
            }
            String ruleName = identifierNode.getInputSpan(grammarSpec);

            ASTNode expressionNode = null;
            for (var child : ruleNode.children()) {
                if (child.label().equals("Expression")) {
                    expressionNode = child;
                    break;
                }
            }
            if (expressionNode == null) {
                throw new IllegalArgumentException("Rule has no Expression child");
            }

            boolean isTransparent = false;
            for (var child : ruleNode.children()) {
                if (child.label().equals(TERMINAL_LABEL) && child.getInputSpan(grammarSpec).equals("~")) {
                    isTransparent = true;
                    break;
                }
                if (child.label().equals("Identifier")) {
                    break;
                }
            }

            Clause clause = buildClause(expressionNode, grammarSpec);

            if (isTransparent) {
                grammarMap.put("~" + ruleName, clause);
            } else {
                grammarMap.put(ruleName, clause);
            }
        }

        // Validate rules
        for (var key : grammarMap.keySet()) {
            if (key.startsWith("~") && grammarMap.containsKey(key.substring(1)) ||
                !key.startsWith("~") && grammarMap.containsKey("~" + key)) {
                throw new IllegalArgumentException("Rule \"" + key + "\" cannot be both transparent and non-transparent");
            }
        }
        for (var clause : grammarMap.values()) {
            clause.checkRuleRefs(grammarMap);
        }

        return grammarMap;
    }

    private static Clause buildClause(ASTNode node, String input) {
        switch (node.label()) {
            case "Expression" -> {
                if (node.children().size() == 1) {
                    return buildClause(node.children().getFirst(), input);
                }
                throw new IllegalArgumentException("Expression should have exactly one child, got " + node.children().size());
            }
            case "Choice" -> {
                var sequences = node.children().stream()
                    .filter(c -> c.label().equals("Sequence"))
                    .map(c -> buildClause(c, input))
                    .toList();
                if (sequences.isEmpty()) {
                    throw new IllegalArgumentException("Choice has no Sequence children");
                }
                return sequences.size() == 1 ? sequences.getFirst() : new First(sequences);
            }
            case "Sequence" -> {
                var prefixes = node.children().stream()
                    .filter(c -> c.label().equals("Prefix"))
                    .map(c -> buildClause(c, input))
                    .toList();
                if (prefixes.isEmpty()) {
                    throw new IllegalArgumentException("Sequence has no Prefix children");
                }
                return prefixes.size() == 1 ? prefixes.getFirst() : new Seq(prefixes);
            }
            case "Prefix" -> {
                String prefixOp = null;
                for (var child : node.children()) {
                    if (child.label().equals(TERMINAL_LABEL)) {
                        String text = child.getInputSpan(input);
                        if (text.equals("&") || text.equals("!") || text.equals("~")) {
                            prefixOp = text;
                            break;
                        }
                    }
                }

                ASTNode operand = null;
                for (var child : node.children()) {
                    if (child.label().equals("Prefix") || child.label().equals("Suffix")) {
                        operand = child;
                        break;
                    }
                }
                if (operand == null) {
                    throw new IllegalArgumentException("Prefix has no Prefix/Suffix child");
                }
                Clause operandClause = buildClause(operand, input);

                return switch (prefixOp) {
                    case "&" -> new FollowedBy(operandClause);
                    case "!" -> new NotFollowedBy(operandClause);
                    case "~" -> operandClause;
                    case null -> operandClause;
                    default -> operandClause;
                };
            }
            case "Suffix" -> {
                String suffixOp = null;
                for (var child : node.children()) {
                    if (child.label().equals(TERMINAL_LABEL)) {
                        String text = child.getInputSpan(input);
                        if (text.equals("*") || text.equals("+") || text.equals("?")) {
                            suffixOp = text;
                            break;
                        }
                    }
                }

                ASTNode operand = null;
                for (var child : node.children()) {
                    if (child.label().equals("Suffix") || child.label().equals("Primary")) {
                        operand = child;
                        break;
                    }
                }
                if (operand == null) {
                    throw new IllegalArgumentException("Suffix has no Suffix/Primary child");
                }
                Clause operandClause = buildClause(operand, input);

                return switch (suffixOp) {
                    case "*" -> new ZeroOrMore(operandClause);
                    case "+" -> new OneOrMore(operandClause);
                    case "?" -> new Optional(operandClause);
                    case null -> operandClause;
                    default -> operandClause;
                };
            }
            case "Primary" -> {
                for (var child : node.children()) {
                    if (!child.label().equals(TERMINAL_LABEL)) {
                        return buildClause(child, input);
                    }
                }
                throw new IllegalArgumentException("Primary has no semantic child");
            }
            case "Parens" -> {
                for (var child : node.children()) {
                    if (child.label().equals("Expression")) {
                        return buildClause(child, input);
                    }
                }
                return new Nothing();
            }
            case "Identifier" -> {
                return new Ref(node.getInputSpan(input));
            }
            case "StringLiteral" -> {
                String text = node.getInputSpan(input);
                return new Str(unescapeString(text.substring(1, text.length() - 1)));
            }
            case "CharLiteral" -> {
                String text = node.getInputSpan(input);
                return new Char(unescapeChar(text.substring(1, text.length() - 1)));
            }
            case "CharClass" -> {
                return buildCharClass(node, input);
            }
            case "AnyChar" -> {
                return new AnyChar();
            }
            default -> throw new IllegalArgumentException("Unknown AST node label: " + node.label());
        }
    }

    private static Clause buildCharClass(ASTNode node, String input) {
        boolean negated = false;
        for (var child : node.children()) {
            if (child.label().equals(TERMINAL_LABEL) && child.getInputSpan(input).equals("^")) {
                negated = true;
                break;
            }
        }

        List<int[]> ranges = new ArrayList<>();
        for (var child : node.children()) {
            if (child.label().equals("CharRange")) {
                var charClassChars = child.children().stream()
                    .filter(c -> c.label().equals("CharClassChar"))
                    .toList();
                if (charClassChars.size() != 2) {
                    throw new IllegalArgumentException("CharRange must have exactly 2 CharClassChar children");
                }
                String lo = extractCharClassCharValue(charClassChars.get(0), input);
                String hi = extractCharClassCharValue(charClassChars.get(1), input);
                ranges.add(new int[]{lo.codePointAt(0), hi.codePointAt(0)});
            } else if (child.label().equals("CharClassChar")) {
                String ch = extractCharClassCharValue(child, input);
                int cp = ch.codePointAt(0);
                ranges.add(new int[]{cp, cp});
            }
        }

        if (ranges.isEmpty()) {
            throw new IllegalArgumentException("CharClass has no character items");
        }

        return new CharSet(ranges, negated);
    }

    private static String extractCharClassCharValue(ASTNode node, String input) {
        for (var child : node.children()) {
            if (child.label().equals("EscapeSequence")) {
                return unescapeChar(child.getInputSpan(input));
            }
        }
        return unescapeChar(node.getInputSpan(input));
    }
}

File: python/squirrelparser/meta_grammar.py
"""MetaGrammar: A grammar for defining PEG grammars."""

from __future__ import annotations

from .clause import Clause
from .combinators import (
    First,
    FollowedBy,
    NotFollowedBy,
    OneOrMore,
    Optional,
    Ref,
    Seq,
    ZeroOrMore,
)
from .terminals import AnyChar, Char, CharSet, Nothing, Str
from .parser import Parser
from .cst_node import ASTNode, build_ast
from .utils import unescape_char, unescape_string

_TERMINAL_LABEL = '<Terminal>'


class MetaGrammar:
    """MetaGrammar: A grammar for defining PEG grammars."""

    # The meta-grammar rules for parsing PEG grammars.
    rules: dict[str, Clause] = {
        'Grammar': Seq([Ref('WS'), OneOrMore(Seq([Ref('Rule'), Ref('WS')]))]),
        'Rule': Seq([
            Optional(Str('~')),
            Ref('Identifier'),
            Ref('WS'),
            Str('<-'),
            Ref('WS'),
            Ref('Expression'),
            Ref('WS'),
            Str(';'),
            Ref('WS'),
        ]),
        'Expression': Ref('Choice'),
        'Choice': Seq([
            Ref('Sequence'),
            ZeroOrMore(Seq([Ref('WS'), Str('/'), Ref('WS'), Ref('Sequence')])),
        ]),
        'Sequence': Seq([Ref('Prefix'), ZeroOrMore(Seq([Ref('WS'), Ref('Prefix')]))]),
        'Prefix': First([
            Seq([Str('&'), Ref('WS'), Ref('Prefix')]),
            Seq([Str('!'), Ref('WS'), Ref('Prefix')]),
            Seq([Str('~'), Ref('WS'), Ref('Prefix')]),
            Ref('Suffix'),
        ]),
        'Suffix': First([
            Seq([Ref('Suffix'), Ref('WS'), Str('*')]),
            Seq([Ref('Suffix'), Ref('WS'), Str('+')]),
            Seq([Ref('Suffix'), Ref('WS'), Str('?')]),
            Ref('Primary'),
        ]),
        'Primary': First([
            Ref('Identifier'),
            Ref('StringLiteral'),
            Ref('CharLiteral'),
            Ref('CharClass'),
            Ref('AnyChar'),
            Ref('Parens'),
        ]),
        'Parens': Seq([Str('('), Ref('WS'), Optional(Ref('Expression')), Ref('WS'), Str(')')]),
        'Identifier': Seq([
            First([CharSet.range('a', 'z'), CharSet.range('A', 'Z'), Char('_')]),
            ZeroOrMore(First([
                CharSet.range('a', 'z'),
                CharSet.range('A', 'Z'),
                CharSet.range('0', '9'),
                Char('_'),
            ])),
        ]),
        'StringLiteral': Seq([
            Str('"'),
            ZeroOrMore(First([
                Ref('EscapeSequence'),
                Seq([NotFollowedBy(First([Str('"'), Str('\\')])), AnyChar()]),
            ])),
            Str('"'),
        ]),
        'CharLiteral': Seq([
            Str("'"),
            First([
                Ref('EscapeSequence'),
                Seq([NotFollowedBy(First([Str("'"), Str('\\')])), AnyChar()]),
            ]),
            Str("'"),
        ]),
        'EscapeSequence': Seq([
            Str('\\'),
            First([
                Char('n'),
                Char('r'),
                Char('t'),
                Char('\\'),
                Char('"'),
                Char("'"),
                Char('['),
                Char(']'),
                Char('-'),
            ]),
        ]),
        'CharClass': Seq([
            Str('['),
            Optional(Str('^')),
            OneOrMore(First([Ref('CharRange'), Ref('CharClassChar')])),
            Str(']'),
        ]),
        'CharRange': Seq([Ref('CharClassChar'), Str('-'), Ref('CharClassChar')]),
        'CharClassChar': First([
            Ref('EscapeSequence'),
            Seq([NotFollowedBy(First([Str(']'), Str('\\'), Str('-')])), AnyChar()]),
        ]),
        'AnyChar': Str('.'),
        '~WS': ZeroOrMore(First([
            Char(' '),
            Char('\t'),
            Char('\n'),
            Char('\r'),
            Ref('Comment'),
        ])),
        'Comment': Seq([
            Str('#'),
            ZeroOrMore(Seq([NotFollowedBy(Char('\n')), AnyChar()])),
            Optional(Char('\n')),
        ]),
    }

    @classmethod
    def parse_grammar(cls, grammar_spec: str) -> dict[str, Clause]:
        """Parse a grammar specification and return the rules."""
        parser = Parser(rules=cls.rules, top_rule_name='Grammar', input=grammar_spec)
        parse_result = parser.parse()
        if parse_result.has_syntax_errors:
            errors = [str(e) for e in parse_result.get_syntax_errors()]
            raise ValueError(f"Failed to parse grammar. Syntax errors:\n{chr(10).join(errors)}")

        ast = build_ast(parse_result)
        grammar_map: dict[str, Clause] = {}

        for rule_node in ast.children:
            if rule_node.label != 'Rule':
                continue

            identifier_node = next((c for c in rule_node.children if c.label == 'Identifier'), None)
            if identifier_node is None:
                raise ValueError('Rule has no Identifier child')
            rule_name = identifier_node.get_input_span(grammar_spec)

            expression_node = next((c for c in rule_node.children if c.label == 'Expression'), None)
            if expression_node is None:
                raise ValueError('Rule has no Expression child')

            is_transparent = False
            for child in rule_node.children:
                if child.label == _TERMINAL_LABEL and child.get_input_span(grammar_spec) == '~':
                    is_transparent = True
                    break
                if child.label == 'Identifier':
                    break

            clause = _build_clause(expression_node, grammar_spec)

            if is_transparent:
                grammar_map[f'~{rule_name}'] = clause
            else:
                grammar_map[rule_name] = clause

        # Validate rules
        for key in grammar_map:
            if ((key.startswith('~') and key[1:] in grammar_map) or
                (not key.startswith('~') and f'~{key}' in grammar_map)):
                raise ValueError(f'Rule "{key}" cannot be both transparent and non-transparent')
        for clause in grammar_map.values():
            clause.check_rule_refs(grammar_map)

        return grammar_map


# ------------------------------------------------------------------------------------------------------------------


def _build_clause(node: ASTNode, input_str: str) -> Clause:
    match node.label:
        case 'Expression':
            if len(node.children) == 1:
                return _build_clause(node.children[0], input_str)
            raise ValueError(f'Expression should have exactly one child, got {len(node.children)}')

        case 'Choice':
            sequences = [_build_clause(c, input_str) for c in node.children if c.label == 'Sequence']
            if not sequences:
                raise ValueError('Choice has no Sequence children')
            return sequences[0] if len(sequences) == 1 else First(sequences)

        case 'Sequence':
            prefixes = [_build_clause(c, input_str) for c in node.children if c.label == 'Prefix']
            if not prefixes:
                raise ValueError('Sequence has no Prefix children')
            return prefixes[0] if len(prefixes) == 1 else Seq(prefixes)

        case 'Prefix':
            prefix_op = None
            for child in node.children:
                if child.label == _TERMINAL_LABEL:
                    text = child.get_input_span(input_str)
                    if text in ('&', '!', '~'):
                        prefix_op = text
                        break

            operand = next((c for c in node.children if c.label in ('Prefix', 'Suffix')), None)
            if operand is None:
                raise ValueError('Prefix has no Prefix/Suffix child')
            operand_clause = _build_clause(operand, input_str)

            match prefix_op:
                case '&':
                    return FollowedBy(operand_clause)
                case '!':
                    return NotFollowedBy(operand_clause)
                case '~':
                    return operand_clause
                case _:
                    return operand_clause

        case 'Suffix':
            suffix_op = None
            for child in node.children:
                if child.label == _TERMINAL_LABEL:
                    text = child.get_input_span(input_str)
                    if text in ('*', '+', '?'):
                        suffix_op = text
                        break

            operand = next((c for c in node.children if c.label in ('Suffix', 'Primary')), None)
            if operand is None:
                raise ValueError('Suffix has no Suffix/Primary child')
            operand_clause = _build_clause(operand, input_str)

            match suffix_op:
                case '*':
                    return ZeroOrMore(operand_clause)
                case '+':
                    return OneOrMore(operand_clause)
                case '?':
                    return Optional(operand_clause)
                case _:
                    return operand_clause

        case 'Primary':
            for child in node.children:
                if child.label != _TERMINAL_LABEL:
                    return _build_clause(child, input_str)
            raise ValueError('Primary has no semantic child')

        case 'Parens':
            for child in node.children:
                if child.label == 'Expression':
                    return _build_clause(child, input_str)
            return Nothing()

        case 'Identifier':
            return Ref(node.get_input_span(input_str))

        case 'StringLiteral':
            text = node.get_input_span(input_str)
            return Str(unescape_string(text[1:-1]))

        case 'CharLiteral':
            text = node.get_input_span(input_str)
            return Char(unescape_char(text[1:-1]))

        case 'CharClass':
            return _build_char_class(node, input_str)

        case 'AnyChar':
            return AnyChar()

        case _:
            raise ValueError(f'Unknown AST node label: {node.label}')


# ------------------------------------------------------------------------------------------------------------------


def _build_char_class(node: ASTNode, input_str: str) -> Clause:
    negated = False
    for child in node.children:
        if child.label == _TERMINAL_LABEL and child.get_input_span(input_str) == '^':
            negated = True
            break

    ranges: list[tuple[int, int]] = []
    for child in node.children:
        if child.label == 'CharRange':
            char_class_chars = [c for c in child.children if c.label == 'CharClassChar']
            if len(char_class_chars) != 2:
                raise ValueError('CharRange must have exactly 2 CharClassChar children')
            lo = _extract_char_class_char_value(char_class_chars[0], input_str)
            hi = _extract_char_class_char_value(char_class_chars[1], input_str)
            ranges.append((ord(lo), ord(hi)))
        elif child.label == 'CharClassChar':
            ch = _extract_char_class_char_value(child, input_str)
            cp = ord(ch)
            ranges.append((cp, cp))

    if not ranges:
        raise ValueError('CharClass has no character items')

    return CharSet(ranges, inverted=negated)


# ------------------------------------------------------------------------------------------------------------------


def _extract_char_class_char_value(node: ASTNode, input_str: str) -> str:
    for child in node.children:
        if child.label == 'EscapeSequence':
            return unescape_char(child.get_input_span(input_str))
    return unescape_char(node.get_input_span(input_str))
